### PR TITLE
Utility classes for Vaadin/Lumo

### PR DESCRIPTION
**Pull request by anezthes**
Initial attempt at creating reusable CSS classes using Lumo's CSS custom properties.

Also includes classes for creating flexbox-based layouts.

### CODE REVIEW

**Comment by jcgueriaud1** [comment]
Is it possible to add the theme rtl for margin left and right?

**Comment by jouni** [comment]
I wonder if these utility classes would only be part of Vaadin 19+, or would they be backported to Vaadin 14?

If they are only for 19+, we should use CSS logical properties, and not for example `margin-left`.

**Comment by rolfsmeds** [comment]
- Shouldn't we prefix the classnames, e.g. with `lumo-` or `v-`? Especially some of the more generic names like `sticky` and `fixed` are quite likely to conflict with custom classnames.
- Names of certain text color classes, e.g. `header-text`, don't really sounds like color-only styles.
- I'm skeptical about the granularity of the flexbox classes. Is it meaningful to have separate classes for `flex`, `flex-row` and `flex-wrap`, instead of combining them to something more convenient like `flex-row-wrap`?

**Comment by rolfsmeds** [comment]
Also I guess we should have classes like size-full, width-full, height-full (or whatever the appropriate names would be).

Another thing that I often end up doing in flex layouts is setting `flex-shrink:0` (to prevent flex items from shrinking) and `min-width:0`/`min-height:0` (since the default is auto, which means intrinsic size of flex items, which means no shrinking beyond that). Perhaps it would make sense to create some convenient combination classes for these so that I could just apply "flex-column shrink-children" or something like that, instead of applying "min-height-zero" to each child separately...

**Comment by lkraav** [comment]
Since Tailwind has a bunch of mindshare currently, perhaps it's prudent to pick up some tricks directly from their sleeve?

We're still in the process of building our utility class set, so I feel like not enough direct experience yet to make an educated recommendation for @rolfsmeds questions above.

**Comment by anezthes** [comment]
> Is it possible to add the theme rtl for margin left and right?

Added end/start variations.

**Comment by anezthes** [comment]
> * Shouldn't we prefix the classnames, e.g. with `lumo-` or `v-`? Especially some of the more generic names like `sticky` and `fixed` are quite likely to conflict with custom classnames.

I've prefixed all classes with `v-` for now.

> * Names of certain text color classes, e.g. `header-text`, don't really sounds like color-only styles.

Those were taken from the [Lumo demo](https://github.com/vaadin/vaadin-lumo-styles/blob/master/demo/colors.html#L459).

[Boostrap](https://getbootstrap.com/docs/4.0/utilities/colors/#color) and [Tailwind](https://tailwindcss.com/docs/text-color) prefix their color styles with `.text-`.

We could do that, e.g. `.v-text-primary` and `.v-text-secondary`, or use `color`, e.g. `.v-color-primary` and `.v-color-secondary`.

(I took the liberty to use the latter option for now. Let me know what you think.)

> * I'm skeptical about the granularity of the flexbox classes. Is it meaningful to have separate classes for `flex`, `flex-row` and `flex-wrap`, instead of combining them to something more convenient like `flex-row-wrap`?

Also inspired by Boostrap and Tailwind. I used a similar approach in the [Material theme for Vaadin 8](https://github.com/vaadin/material-theme-fw8/blob/master/material-theme-fw8/src/main/resources/VAADIN/themes/material-design/layout/flex-layout.scss) and liked the ease of use and flexibility.

Of course, everybody's different, and I'm definitely open to suggestions. Guess the only thing we need to consider then is if the combinations will grow numerous.

**Comment by marcushellberg** [comment]
I think we should target V19+ with these. I believe they'll be most useful in the short term for Fusion users, although in the long run we may also be able to switch out Flow VL/HL etc implementations with lighter weight divs with the right classes.

**Comment by marcushellberg** [comment]
What do you think about adding classes for gaps? I often want gaps in flex boxes and stupid Safari hasn't implemented them yet. 

Something like 

```css
.gap-m > :not(:last-child) {
  margin-block-end: var(--lumo-space-m);
  margin-inline-end: var(--lumo-space-m);
}
```

**Comment by anezthes** [comment]
> What do you think about adding classes for gaps? I often want gaps in flex boxes and stupid Safari hasn't implemented them yet.

Columns and gaps are coming soon™.

**Comment by rolfsmeds** [comment]
> > What do you think about adding classes for gaps? I often want gaps in flex boxes and stupid Safari hasn't implemented them yet.
> 
> Columns and gaps are coming soon™.

I'm a bit wary of implementing margin-based gaps right now, as they don't work well with wrapping, and we'll want to refactor them to use the `gap` property as soon as it's properly supported, but we may not be able to do that without breaking existing applications if we introduce them in the LTS now.

Some kind of css grid based layout classes with gap support would be another thing though.

**Comment by marcushellberg** [comment]
Yeah, I agree that proper `gap` would be nicer. But it would be super convenient to have something already in the meantime.

**Comment by marcushellberg** [comment]
I copied these over to my project to see what they felt like in action. Overall, I really liked the experience. Saved me from having to write basically any CSS. Some notes and thoughts:

* I didn't like the `v-` prefix to every class name, it felt verbose and a bit redundant. I would vote to get rid of the prefix for simplicity. Tailwind and Bootstrap don't prefix their classes either. Maybe we could add that as a build-time option you can turn on if someone really wants a prefix?
* I missed size helpers. Especially something like full-width, full-height, full-size. 
* Gaps would be nice. 
* I didn't need it right now, but I could see that many apps could use a responsive layouting system/grid like the above mentioned libraries. Maybe that's a separate PR?

**Comment by marcushellberg** [comment]
I'm also missing options for setting margin/padding to 0.

**Comment by jouni** [comment]
When do you need to reset padding to zero?

**Comment by marcushellberg** [comment]
My actual need was for 0 margin, figured someone might need a 0 padding some day and keeping it consistent would make sense.

**Comment by jouni** [comment]
> someone might need a 0 padding some day

Right. I was trying to think when that might happen. IIRC browser defaults do not include any padding. Some of our components do, thought (notably input fields), so I was wondering if that was the reason.

I’m not sure if consistency is a good reason to include unused CSS to the bundle.

**Comment by anezthes** [comment]
**Update**

Added:
* CSS grid with up to 12 columns and 12 rows
* Ability set the column and row spans
* Height and width classes (`xs` through `xl`, with `full` option)
* Icon sizes (`s`, `m` and `l`)
* Spacing based on both `margin` and `gap`

**Comment by anezthes** [comment]
Until it's official (reviewed, accepted and whatnot) you can use this highly volatile package at your own discretion/peril/risk/demise:
https://www.npmjs.com/package/lumo-css-framework

**Comment by rolfsmeds** [comment]
I'll just dump a list of open questions here for future discussion:

- Should all flex-classes include display:flex (why would you not want that?)
- Should all utility classes be in a single stylesheet, or should we split them by type so you can pick which you want to use?
- Should we keep the v- prefix? E.g. [Tailwind](https://tailwindcss.com/docs/flex) is prefixless, while [Ionic](https://ionicframework.com/docs/layout/css-utilities) has `ion-` prefix.
- What about Material theme?
- Do we want to create and maintain our own utility css, or should we just use Tailwind directly, as that's what all the cool kids use today anyway? (We do want the classes to use Lumo properties, but apparently TW can be [customized](https://tailwindcss.com/docs/configuration))
- Should we remove the global element styles in Lumo (e.g. [typography](https://github.com/vaadin/vaadin-lumo-styles/blob/master/typography.html#L98) in favor of utility classes?
- Should the utility classes by a default import of Lumo in Flow / part of [all-imports](https://github.com/vaadin/vaadin-lumo-styles/blob/master/all-imports.html) / loaded by default as part of Lumo by the new [custom theme feature](https://vaadin.com/docs-beta/latest/ds/customization/custom-theme/)? (I would say **no**)

**Comment by anezthes** [comment]
> Should all flex-classes include display:flex (why would you not want that?)

When you want to use `inline-flex`.

> Should all utility classes be in a single stylesheet, or should we split them by type so you can pick which you want to use?

Could split them up, and offer a "import all" option.

> Should we keep the v- prefix? E.g. Tailwind is prefixless, while Ionic has ion- prefix.

Personally, I'd prefer no prefix. I think @marcushellberg preferred that route as well.

> What about Material theme?

Shouldn't take long to modify the typography and white-space (margins, paddings, etc) styles. Could (finally) introduce proper CSS custom properties at the same time, and update them to the latest MD spec.

> Do we want to create and maintain our own utility css, or should we just use Tailwind directly, as that's what all the cool kids use today anyway?

¯\_(ツ)_/¯

> Should we remove the global element styles in Lumo (e.g. typography in favor of utility classes?

Good question. I'd probably defer to @jouni. Might still be good to have some kind of "default"?

> Should the utility classes by a default import of Lumo in Flow

I think opt-in to start, and if it becomes popular, then we could consider auto-importing it in the future.

Just IMO.

**Comment by jcgueriaud1** [comment]
> * Do we want to create and maintain our own utility css, or should we just use Tailwind directly, as that's what all the cool kids use today anyway? (We do want the classes to use Lumo properties, but apparently TW can be [customized](https://tailwindcss.com/docs/configuration))

In my opinion, Tailwind is a good option if you can remove unused CSS. I'm not sure if it's really possible with Fusion+Flow. (I would say it comes with a lot of restrictions). 4Mb of uncompressed css is quite huge.
Tailwind (or these utility classes) does not work in shadow dom.

There are also these kind of project: https://github.com/tw-in-js/twind
It solves the shadow dom issue and tree shaking but:
* It's probably only for super crazy cool kids :D
* I don't think you can use it easily in Flow.

**Comment by jouni** [comment]
> Tailwind is a good option if you can remove unused CSS. I'm not sure if it's really possible with Fusion+Flow. (I would say it comes with a lot of restrictions)

Yeah, that would be a prerequisite to use Tailwind, that the official webpack plugin works with a Vaadin setup as well, and automatically removes unused classes. AFAIK it should be possible, as the plugin basically does it in a very “dumb” way, by checking if a string is found in one of the source files, then it keeps the class in the final build. We could perhaps extend the check to include .java files as well, so that Flow views would be considered as well.

**Comment by jouni** [comment]
> Could (finally) introduce proper CSS custom properties at the same time (for Material)

I’m not super excited about introducing custom properties for Material. Our idea has been to move to the opposite direction, that we base the Material theme on top of Lumo eventually.

**Comment by marcushellberg** [comment]
Using Tailwind is an interesting alternative. It would give us all the functionality and documentation out of the box. The downside, besides having to setup a build is the dependency on tailwind. Will it become a problem in 3 years? Although it seems to be MIT licensed, so worst case we could maintain some fork I guess as a contingency.

File: utility-classes.html [hunk 1]
```diff
@@ -108,6 +108,22 @@
...
+      }
+
+      /* === Flex shrink === */
+      .v-flex-shrink-0 {
+        flex-shrink: 0;
+      }
+      .v-flex-shrink {
+        flex-shrink: 1;
+      }
+
```
Comment (jouni) [inline utility-classes.html:126]: I would expect to have `.v-flex-none` and `.v-flex-auto` available. I think those cover 80–90% of use cases.

Comment (anezthes) [reply utility-classes.html:132]: Added.

**Comment by rolfsmeds** [comment]
Summary of decisions:

- This stylesheet will become an official but opt-in part of Lumo in V20.
- Prefixes will be removed to align with other popular utility class frameworks.
- Classes whose implementation is closely aligned with Tailwind should use the same classnames to ease adoption.
- Classes whose implementation _could_ be aligned with TW without clashes with Lumo or Vaadin aspects should be aligned.
- Classes whose implementation differs significantly from TW (e.g. due to different scales compared to Lumo properties) should have different classnames.
- The non-aligned classnames should follow a similar logic however.
- We will not use Tailwind instead of this for now, but keep the option open to introduce TW as an alternative to this in the future.
- Material theme and Flow API is not included in the current scope.
- Documentation needs to be added to Vaadin Design System / Foundation docs before V20 PiT.
- Let's try to get as much feedback from the community before that.

**Comment by web-padawan** [comment]
Please close this PR and submit a new one to https://github.com/vaadin/vaadin-web-components.

We also need to have an issue with the acceptance criteria that the PR would be linked to.

**Comment by DiegoCardoso** [comment]
Closing this PR.

It was moved to the monorepo and merged here (https://github.com/vaadin/web-components/pull/2121) and it's released since [v21.0.0-alpha9](https://github.com/vaadin/web-components/releases/tag/v21.0.0-alpha9).

**Pull request closed without merging**